### DISCUSSION TITLE
Nintendo Switch: "Registration is Required to Use This Network"

### DISCUSSION THREAD

**Discussion by Klaaktu** [shadowsocks/shadowsocks-rust#491] [Q&A] [answered]
Setting the proxy address to HTTP listening address of `sslocal`(shadowsocks-rust) works fine on my laptop and phone, but on Nintendo Switch an error message is shown when it's checking the connection to Internet
> "Registration is Required to Use This Network."

When I click next then it opens the browser to `ctest.cdn.nintendo.net` which just says "ok" on the page. However when I close the browser Switch just disconnects.
This problem doesn't happen when I use Shadowsocks-Windows as the local proxy. It says "successful" as normal.

**Comment by DuckSoft** [comment, marked as answer]
Is there a [Captive Portal](https://en.wikipedia.org/wiki/Captive_portal) on your WiFi network before actually connecting to the Internet? 

If so, you must have your session (or mac address) acknowledged by the portal, then you can (probably) access your own proxy service.

**Reply by Klaaktu** [reply]
Thanks, but I don't think so... I set up the WiFi myself. It also just connects after turning off proxy.

**Reply by DuckSoft** [reply]
Here are the aggregated results from Wireshark capture:
 
<table>
<tr>
<th>Original</td>
<th colspan=2>Proxied</td>
</tr>
<tr>
<td align=center>no software</td>
<td align=center>v2ray</td>
<td align=center>ss-rust</td>
</tr>
<tr>
<td><img alt="original" src="https://user-images.githubusercontent.com/7822648/114284591-7b99c380-9a83-11eb-850e-f125383c31d6.png"></td>
<td><img alt="v2ray" src="https://user-images.githubusercontent.com/7822648/114284569-658c0300-9a83-11eb-8fb4-28dee121b119.png"></td>
<td><img alt="ssrust" src="https://user-images.githubusercontent.com/7822648/114284508-04fcc600-9a83-11eb-8a63-ec2f880b1e15.png"></td>
</tr>
<tr>
<td align=center>WORKING</td>
<td align=center>WORKING</td>
<td align=center><b>NOT WORKING</b></td>
</tr>
</table>

My guess is that Nintendo Switch depends on a non-RFC compliant handling for Captive Portal detection.
As is stated in RFC 2616 - "Hypertext Transfer Protocol -- HTTP/1.1", Section 4.2, "Message Headers": 
> Each header field consists of a name followed by a colon (":") and the field value. ***Field names are case-insensitive.***

As seen above, Shadowsocks-Rust converted all the header keys to lower cases. To prove my assumptions, I'll write a MITM proxy program, only hijacking the first connection and response with the same content as Shadowsocks-Rust will do, but with the headers cases **returned to normal**. Afterwards, the connection is directly made with Shadowsocks-rust. Let's see what will happen now.

cc @zonyitoo @database64128 @xiaokangwang

**Reply by DuckSoft** [reply]
The experiment is done. Here is the code used for my PoC:

<details>
<summary> Golang Code for Proof of Concept </summary>

```go
package main

import (
	"flag"
	"io"
	"log"
	"net"
)

var (
	ListenAddr = flag.String("listen",  "0.0.0.0:8002", "listen address")
	RemoteAddr = flag.String("remote", "127.0.0.1:8001", "remote address")
)

func main() {
	listener, err := net.Listen("tcp4", *ListenAddr)
	if err != nil {
		panic(err)
	}
	conn, err := listener.Accept()
	if err != nil {
		panic(err)
	}
	log.Printf("conn accepted: %v", conn.RemoteAddr().String())
	b := make([]byte, 4096)
	n, err := io.ReadAtLeast(conn, b, 20)
	if err != nil {
		panic(err)
	}
	log.Printf("received:\n%s", string(b[:n]))
	_, err = conn.Write([]byte("HTTP/1.1 200 OK\r\nContent-Length: 2\r\nExpires: Sat, 10 Apr 2021 21:00:12 GMT\r\nCache-Control: max-age=0, no-cache, no-store\r\nPragma: no-cache\r\nDate: Sat, 10 Apr 2021 21:00:12 GMT\r\nConnection: keep-alive\r\nX-Organization: Nintendo\r\nContent-Type: text/plain\r\n\r\nok"))
	if err != nil {
		panic(err)
	}
	log.Printf("fake response sent")
	conn.Close()
	for {
		conn, err := listener.Accept()
		if err != nil {
			continue
		}
		go func() {
			defer conn.Close()
			remoteConn, err := net.Dial("tcp4", *RemoteAddr)
			if err != nil {
				return
			}
			defer conn.Close()
			go io.Copy(remoteConn, conn)
			io.Copy(conn, remoteConn)
		}()
	}
}
```

</details>

I noticed that after the first HTTP GET proxy request the following requests are all CONNECT requests, which won't have such problem, so I made a toy program to make the first connection speak normal case instead of lower ones. Afterwards, the connections are forwarded as-is.

The outcome confirms that Nintendo Switch is case-sensitive about the keys of Header in Captive Portal detection response of HTTP proxy. I think this should be reported to Nintendo relevant personnel as a bug, rather than a bug in Shadowsocks-Rust. After all, the RFC said headers don't need to be case sensitive. 

Let's wait @zonyitoo for the final decision.

**Reply by DuckSoft** [reply]
I narrowed down to the final results! It turns out that if this `X-Organization: Nintendo` header key `X-Organization` get lower cased (or in any other case you like, for example, `x-OrGaNiZaTiON`), Nintendo Switch will think that this HTTP proxy isn't working!

**Reply by DuckSoft** [reply]
I've filed a ticket to Nintendo reporting this issue. Now let's look at the code of our Shadowsocks-Rust and see what we can do to workaround the situation.

**Reply by DuckSoft** [reply]
https://docs.rs/hyper/0.14.5/hyper/client/struct.Builder.html#method.http1_title_case_headers
Got this. Working on a patch...

**Reply by DuckSoft** [reply]
Uh, I should have searched...
It's already put by @zonyitoo: https://github.com/hyperium/hyper/issues/2313

**Reply by DuckSoft** [reply]
@klaaktu I've made a workaround in https://github.com/shadowsocks/shadowsocks-rust/actions/runs/736992309.
You may build commit 0151f6ac712cea4491d3bf98b25ccab2dd0d4ffb locally to use now, or you can wait for the stable release.

Case closed!

**Reply by Klaaktu** [reply]
Thanks!
(Oh wow didn't know the issue already exists. Just got a bunch of generic stuff when I tried googling...)

**Reply by zonyitoo** [reply]
@DuckSoft Where did you send ticket to Nintendo? I found it long time ago and couldn't find a way to tell them about their stupid code.

**Reply by DuckSoft** [reply]
Nintendo Support (Japanese): https://support.nintendo.co.jp/

---
ps: Some update (2021-04-13 10:49 am) from Nintendo:

```
このたびはお手数をおかけしまして、誠に申し訳ございません。

お客様より頂戴いたしましたご意見・ご要望は、関連部門に報告させていただきます。

今後とも弊社製品をご愛顧賜りますようお願い申しあげます。
```

**Comment by zonyitoo** [comment]
Yes. It is a known issue. It is because [`hyper`](https://crates.io/crates/hyper), the HTTP library supporting `sslocal`'s HTTP proxy, doesn't preserve header's case. `hyper` will convert all headers' key to lower case.

HTTP headers are case insensitive, so what `hyper` does is totally correct based on RFC.

But Nintendo Switch's connection test will try to send a `GET` request to `http://ctest.cdn.nintendo.net/` and expecting a response with header `X-Organization: Nintendo`, in case sensitive way.

I opened [an issue](https://github.com/hyperium/hyper/issues/2313) log time ago and they recently found a solution.

**Reply by zonyitoo** [reply]
Oh, there is a **View more** button. I saw you guys have found my issue already.

**Comment by zonyitoo** [comment]
https://github.com/shadowsocks/shadowsocks-rust/commit/665027a708327247e351c7e414c87e579d124689

```
curl 'http://ctest.cdn.nintendo.net/' --proxy 127.0.0.1:1300 -v -H 'User-Agent: NX NIFM/00'
*   Trying 127.0.0.1:1300...
* Connected to 127.0.0.1 (127.0.0.1) port 1300 (#0)
> GET http://ctest.cdn.nintendo.net/ HTTP/1.1
> Host: ctest.cdn.nintendo.net
> Accept: */*
> Proxy-Connection: Keep-Alive
> User-Agent: NX NIFM/00
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Cache-Control: max-age=0, no-cache, no-store
< Content-Length: 2
< Content-Type: text/plain
< Date: Fri, 23 Apr 2021 02:22:21 GMT
< Expires: Fri, 23 Apr 2021 02:22:21 GMT
< Pragma: no-cache
< X-Cache: MISS from SZ-SQUIDWEB-70
< X-Cache-Lookup: MISS from SZ-SQUIDWEB-70:8080
< X-Organization: Nintendo
<
* Connection #0 to host 127.0.0.1 left intact
ok
```

**Reply by Klaaktu** [reply]
Nice!
